### PR TITLE
Fix potential endless loop during merge

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
In case when `merge_max_block_size_bytes` is small enough and tables contain wide rows (strings or tuples) background merges may stuck in an endless loop. This behaviour is fixed. Follow-up for https://github.com/ClickHouse/ClickHouse/pull/59340

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit b4bd7e2f75df3d5eea3e8e8d78d2448c2b32a413 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59812/b4bd7e2f75df3d5eea3e8e8d78d2448c2b32a413/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59812/b4bd7e2f75df3d5eea3e8e8d78d2448c2b32a413/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Processors/Transforms/ColumnGathererTransform.h [hunk 1]
```diff
@@ -191,7 +195,7 @@ void ColumnGathererStream::gather(Column & column_res)
         }
 
         source.pos += len;
-    }
```
Comment (CurtizJ) [inline src/Processors/Transforms/ColumnGathererTransform.h:194]: It's worth to change `allocatedBytes` to `bytes` like it was suggested in https://github.com/ClickHouse/ClickHouse/pull/59340#discussion_r1470354144. It appeared to be more important as it seemed. And also change this in https://github.com/ClickHouse/ClickHouse/blob/5a024dfc0936e062770d0cfaad0805b57c1fba17/src/Processors/Merges/Algorithms/MergedData.h#L125.

Comment (nikitamikhaylov) [reply src/Processors/Transforms/ColumnGathererTransform.h:194]: Done.

**Review by CurtizJ — APPROVED** [review]

**Comment by clickgapai** [comment]
Hi @nikitamikhaylov @CurtizJ — while reviewing this PR I found the following:

- **Coverage gap:** vertical merge endless-loop fix (`ColumnGathererStream::gather` do/while + `byteSize`) has zero regression coverage — https://github.com/ClickHouse/ClickHouse/pull/103841

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**